File: Sources/FWObjC/Module/Controller/TableViewController.m
//
//  TableViewController.m
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

#import "TableViewController.h"
#import <objc/runtime.h>

#if FWMacroSPM

@interface UIView ()

- (NSArray<NSLayoutConstraint *> *)__fw_pinEdgesToSuperview:(UIEdgeInsets)insets;

@end

#else

#import <FWFramework/FWFramework-Swift.h>

#endif

#pragma mark - __FWViewControllerManager+__FWTableViewController

@implementation __FWViewControllerManager (__FWTableViewController)

+ (void)load
{
    __FWViewControllerIntercepter *intercepter = [[__FWViewControllerIntercepter alloc] init];
    intercepter.viewDidLoadIntercepter = @selector(tableViewControllerViewDidLoad:);
    intercepter.forwardSelectors = @{
        @"tableView" : @"__fw_tableView",
        @"tableData" : @"__fw_tableData",
        @"setupTableStyle" : @"__fw_setupTableStyle",
        @"setupTableLayout" : @"__fw_setupTableLayout",
    };
    [[__FWViewControllerManager sharedInstance] registerProtocol:@protocol(__FWTableViewController) withIntercepter:intercepter];
}

- (void)tableViewControllerViewDidLoad:(UIViewController<__FWTableViewController> *)viewController
{
    UITableView *tableView = [viewController tableView];
    tableView.dataSource = viewController;
    tableView.delegate = viewController;
    [viewController.view addSubview:tableView];
    
    if (self.hookTableViewController) {
        self.hookTableViewController(viewController);
    }
    
    if ([viewController respondsToSelector:@selector(setupTableView)]) {
        [viewController setupTableView];
    }
    
    [viewController setupTableLayout];
    [tableView setNeedsLayout];
    [tableView layoutIfNeeded];
}

@end

#pragma mark - UIViewController+__FWTableViewController

@interface UIViewController (__FWTableViewController)

@end

@implementation UIViewController (__FWTableViewController)

- (UITableView *)__fw_tableView
{
    UITableView *tableView = objc_getAssociatedObject(self, _cmd);
    if (!tableView) {
        UITableViewStyle tableStyle = [(id<__FWTableViewController>)self setupTableStyle];
        tableView = [[UITableView alloc] initWithFrame:CGRectZero style:tableStyle];
        tableView.showsVerticalScrollIndicator = NO;
        tableView.showsHorizontalScrollIndicator = NO;
        tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
        if (@available(iOS 15.0, *)) {
            tableView.sectionHeaderTopPadding = 0;
        }
        objc_setAssociatedObject(self, _cmd, tableView, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    }
    return tableView;
}

- (NSMutableArray *)__fw_tableData
{
    NSMutableArray *tableData = objc_getAssociatedObject(self, _cmd);
    if (!tableData) {
        tableData = [[NSMutableArray alloc] init];
        objc_setAssociatedObject(self, _cmd, tableData, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    }
    return tableData;
}

- (UITableViewStyle)__fw_setupTableStyle
{
    return UITableViewStylePlain;
}

- (void)__fw_setupTableLayout
{
    UITableView *tableView = [(id<__FWTableViewController>)self tableView];
    [tableView __fw_pinEdgesToSuperview:UIEdgeInsetsZero];
}

@end

File: Sources/FWFramework/Module/Controller/TableViewController.swift
//
//  TableViewController.swift
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

import UIKit

// MARK: - TableViewControllerProtocol
/// 表格视图控制器协议，可覆写
public protocol TableViewControllerProtocol: ViewControllerProtocol, UITableViewDataSource, UITableViewDelegate {
    
    /// 关联表格数据元素类型，默认Any
    associatedtype TableElement = Any
    
    /// 表格视图，默认不显示滚动条，Footer为空视图。Plain有悬停，Group无悬停
    var tableView: UITableView { get }

    /// 表格数据，默认空数组，延迟加载
    var tableData: [TableElement] { get set }

    /// 渲染表格视图样式，默认Plain
    func setupTableStyle() -> UITableView.Style

    /// 渲染表格视图，setupSubviews之前调用，默认空实现
    func setupTableView()

    /// 渲染表格视图布局，setupSubviews之前调用，默认铺满
    func setupTableLayout()
    
}

extension TableViewControllerProtocol where Self: UIViewController {
    
    /// 表格视图，默认不显示滚动条，Footer为空视图。Plain有悬停，Group无悬停
    public var tableView: UITableView {
        if let result = fw_property(forName: "tableView") as? UITableView {
            return result
        } else {
            let result = UITableView(frame: .zero, style: setupTableStyle())
            result.showsVerticalScrollIndicator = false
            result.showsHorizontalScrollIndicator = false
            result.tableFooterView = UIView(frame: .zero)
            if #available(iOS 15.0, *) {
                result.sectionHeaderTopPadding = 0
            }
            fw_setProperty(result, forName: "tableView")
            return result
        }
    }
    
    /// 表格视图代理，调用时自动生效
    public var tableDelegate: TableViewDelegate {
        return tableView.fw_delegate
    }
    
    /// 表格数据，默认空数组，延迟加载
    public var tableData: [TableElement] {
        get { return fw_property(forName: "tableData") as? [TableElement] ?? [] }
        set { fw_setProperty(newValue, forName: "tableData") }
    }
    
    /// 渲染表格视图样式，默认Plain
    public func setupTableStyle() -> UITableView.Style {
        return .plain
    }
    
    /// 渲染表格视图，setupSubviews之前调用，默认空实现
    public func setupTableView() {}
    
    /// 渲染表格视图布局，setupSubviews之前调用，默认铺满
    public func setupTableLayout() {
        tableView.fw_pinEdges()
    }
    
}

// MARK: - ViewControllerManager+TableViewControllerProtocol
internal extension ViewControllerManager {
    
    func tableViewControllerViewDidLoad(_ viewController: UIViewController) {
        guard let viewController = viewController as? any UIViewController & TableViewControllerProtocol else { return }
        
        let tableView = viewController.tableView
        tableView.dataSource = viewController
        tableView.delegate = viewController
        viewController.view.addSubview(tableView)
        
        hookTableViewController?(viewController)
        
        viewController.setupTableView()
        viewController.setupTableLayout()
        tableView.setNeedsLayout()
        tableView.layoutIfNeeded()
    }
    
}
